crates/nu-command/src/filesystem/ls.rs

```diff
@@ -979,10 +944,32 @@ mod windows_helper {
 #[allow(clippy::type_complexity)]
 fn read_dir(
     f: &Path,
+    span: Span,
 ) -> Result<Box<dyn Iterator<Item = Result<PathBuf, ShellError>> + Send>, ShellError> {
-    let iter = f.read_dir()?.map(|d| {
-        d.map(|r| r.path())
-            .map_err(|e| ShellError::IOError { msg: e.to_string() })
-    });
+    let iter = f
+        .read_dir()
+        .map_err(|error| {
+            if error.kind() == std::io::ErrorKind::PermissionDenied {
+                #[cfg(unix)]
+                let error_msg = format!(
+                    "The permissions of {:o} do not allow access for this user",
+                    f.metadata()
+                        .map(|m| m.permissions().mode() & 0o0777)
+                        .expect("this shouldn't be called since we already know there is a dir")
```

Comment:
I think there may be a minor TOCTOU risk of this panicking if the directory is deleted between the raising of the error and hitting `File::metadata` which runs a `statx()` syscall on the file descriptor. 

Not quite sure how I feel about showing the octal mode repr here. Certainly more informative than bare permission denied but at the same time somewhat cryptic for folks not brought up with old `chmod`. And also not sure if it can properly represent dernied permissions due to ACLs.